File: TableOfContentsSample/Classes/TOCViewController.m
// Copyright 2012 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

#import "TOCViewController.h"
#import "LayerInfo.h"
#import "LayerInfoCell.h"
#import "LayerLegendCell.h"


@interface TOCViewController() <LayerInfoCellDelegate, UITableViewDelegate, UITableViewDataSource>


//mapView to create the TOC for
@property (nonatomic, weak) AGSMapView *mapView;

//this is the map level layer info object, acting as the invisible root of the entire tree. 
@property (nonatomic, strong) LayerInfo *mapViewLevelLayerInfo;

//navbar to control the view
@property (nonatomic, strong) IBOutlet UINavigationBar *navBar;

//tableview to display the layers.
@property (nonatomic, strong) IBOutlet UITableView *tableView;

//adds the layer to the array of layer views to be displayed in the Table of Contents.
//This is called when the layer is loaded.
- (void)insertLayer:(AGSLayer*)layer atIndex:(int)index;

//method to add/remove the layers to/from the ToC
- (void)processMapLayers;

//to dismiss the view. 
- (IBAction)doneButtonPressed;

@end


@implementation TOCViewController

@synthesize mapView = _mapView;
@synthesize mapViewLevelLayerInfo = _mapViewLevelLayerInfo;
@synthesize popOverController = _popOverController;
@synthesize navBar = _navBar;
@synthesize tableView = _tableView;

- (id)initWithMapView:(AGSMapView *)mapView
{
    self = [super initWithNibName:@"TOCViewController" bundle:nil];
    if (self) {
        
        //assign the mapView
        self.mapView = mapView;
        
        //process the map layers. 
        [self processMapLayers];
        
        //create the map level layer info object with id of -2 and layer view as nil.       
        self.mapViewLevelLayerInfo = [[LayerInfo alloc] initWithLayer:nil layerID:-2 name:@"Map" target:nil];
    }
    return self;
}

- (void)didReceiveMemoryWarning
{
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

#pragma mark - View lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
	  
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self processMapLayers];
}


- (void)viewDidUnload
{
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

- (void)dealloc
{
    self.popOverController = nil;
}

#pragma mark -
#pragma mark LayerInfoCellDelegate

- (void)layerVisibilityChanged:(BOOL)visibility forCell:(UITableViewCell *)cell
{
    //retrieve the corresponding cell
    NSIndexPath *cellIndexPath = [self.tableView indexPathForCell:cell];
    
    //get the layer info represented by the cell
    LayerInfo *layerInfo = [[self.mapViewLevelLayerInfo flattenElementsWithCacheRefresh:NO withLegend:YES] objectAtIndex:(cellIndexPath.row)];
    
    //set the visibility of the layer info. 
    [layerInfo setVisible:visibility];
}

#pragma mark -
#pragma mark Helper Methods

- (void)insertLayer:(AGSLayer*)layer atIndex:(int)index
{
    //creates the layer info node for a layer in the map after it is loaded in the map view. and the delegate method is called after the tree is constructed. 
     LayerInfo *layerInfo = [[LayerInfo alloc] initWithLayer:layer layerID:-1 name:layer.name target:self];
    [self.mapViewLevelLayerInfo insertChild:layerInfo atIndex:index];
}

- (void)doneButtonPressed {
    if([[AGSDevice currentDevice] isIPad])
		[self.popOverController dismissPopoverAnimated:YES];
	else
		[self dismissModalViewControllerAnimated:YES];
    
}


#pragma mark -
#pragma mark Table view data source

// Customize the number of sections in the table view.
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    //uses the descendant count with legend. 
    return [self.mapViewLevelLayerInfo descendantAndLegendCount];
}


// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *LayerInfoCellIdentifier = @"LayerInfoCell";
    static NSString *LayerLegendCellIdentifier = @"LayerLegendCell";
    
    LayerInfoCell *layerInfoCell;
    LayerLegendCell *layerLegendCell;
    
    //obtain the object contained in the flat array of the tree nodes, that includes the legend. 
    id tempObject = [[self.mapViewLevelLayerInfo flattenElementsWithCacheRefresh:NO withLegend:YES] objectAtIndex:indexPath.row];
    
    //if it is a legend info class, form a legend info cell. 
    if([tempObject isKindOfClass:[LayerInfo class]])
    {
        LayerInfo *layerInfo = (LayerInfo *)tempObject;    
        layerInfoCell = [[LayerInfoCell alloc] initWithStyle:UITableViewCellStyleDefault
                                              reuseIdentifier:LayerInfoCellIdentifier 
                                                        level:[layerInfo levelDepth] - 1 
                                          canChangeVisibility:layerInfo.canChangeVisibility
                                                   visibility:layerInfo.visible
                                                     expanded:layerInfo.inclusive];   
        
        //assign the title. 
        layerInfoCell.valueLabel.text = layerInfo.layerName;
        
        //assign the delegate to call the method when the visibility is changed. 
        layerInfoCell.layerInfoCellDelegate = self;
        
        //no selection style. 
        layerInfoCell.selectionStyle = UITableViewCellSelectionStyleNone;
        
        return layerInfoCell;
    }
    else
    {
        LegendElement *legentElement = (LegendElement *)tempObject;  
        layerLegendCell = [[LayerLegendCell alloc] initWithStyle:UITableViewCellStyleDefault
                                                  reuseIdentifier:LayerLegendCellIdentifier 
                                                            level:legentElement.level];
        
        //assign the title. 
        layerLegendCell.legendLabel.text = legentElement.title;
        
        //assign the image. 
        layerLegendCell.legendImage.image = legentElement.swatch;
        
        //no selection style. 
        layerLegendCell.selectionStyle = UITableViewCellSelectionStyleNone;
        
        return layerLegendCell;
    } 
    
}



#pragma mark -
#pragma mark Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    //obtain the object contained in the flat array of the tree nodes, that includes the legend. 
	id tempObject = [[self.mapViewLevelLayerInfo flattenElementsWithCacheRefresh:NO withLegend:YES] objectAtIndex:indexPath.row];
    
    //do nothing if this is a legend cell. 
    if([tempObject isKindOfClass:[LegendElement class]])
        return;
    
    //else expand or contract the layer. 
    LayerInfo *layerInfo = (LayerInfo *)tempObject;    
	layerInfo.inclusive = !layerInfo.inclusive;	
	[self.mapViewLevelLayerInfo flattenElementsWithCacheRefresh:YES withLegend:YES];
    
    //reload the table. 
	[tableView reloadData];
}

#pragma mark -
#pragma mark Helper Methods

- (void)processMapLayers
{
    //pruning the tree to remove the layers that are not in the mapview anymore.       
    NSMutableArray *layersToBeRemoved = [NSMutableArray array];
    for (LayerInfo *layerInfo in self.mapViewLevelLayerInfo.children)
    {       
        //bool indicating whether the layer is present or not. 
        BOOL layerPresent = NO;
        
        //iterate through the map layers to check the layer has been removed since previous display. 
        for (AGSLayer *layer in self.mapView.mapLayers) {
            if([layer.name isEqualToString:layerInfo.layerName])
            {
                //layer is present
                layerPresent = YES;
                break;
            }               
        }
        if(layerPresent)
        {
            //if layer is present continue to the next one. 
            continue;
        }
        else
        {
            //if the layer is not present, then add the corresponding layerinfo to the array of layers to be removed. 
            [layersToBeRemoved addObject:layerInfo]; 
        }
    }   
    
    //iterating through the "to be removed" layers, if any, as they were removed from the mapview. 
    for (LayerInfo *layerInfo in layersToBeRemoved)
    {
        //remove the layerinfo.
        [self.mapViewLevelLayerInfo removeChildWithLayerName:layerInfo.layerName];
    }

    
    //adds the layers to the tree ( (starting from top most layer, going to bottom)
    for (int i = 0, j=self.mapView.mapLayers.count-1; i < self.mapView.mapLayers.count ; i++,j--)
    {     
        //get the agsLayer from the mapLayers array 
        AGSLayer *agsLayer = [self.mapView.mapLayers objectAtIndex:j];
        
        //check to see if the agsLayer exists in the toc or not. 
        if(![self.mapViewLevelLayerInfo containsChildWithLayerName:agsLayer.name])
        {
            //if not, check for whether it has been loaded
            if(agsLayer.loaded)
            {            
                //if loaded, insert the layer's 'LayerInfo' at the appropriate index of the ToC
                [self insertLayer:agsLayer atIndex:i];
            }
            else
            {
                //if not loaded yet, add a kvo on the loaded property of the agsLayer. 
                [agsLayer addObserver:self forKeyPath:@"loaded" options:0 context:(__bridge void *)([NSNumber numberWithInt:i])];
            }
        }
    }  
    
    //refresh the flattened tree cache. 
    [self.mapViewLevelLayerInfo flattenElementsWithCacheRefresh:YES withLegend:YES];
      
    //reload the table view. 
    [self.tableView reloadData];
    
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context 
{    
    AGSLayer *layer = (AGSLayer *)object;
    
    //obtain the index from the context
    NSNumber *index = (__bridge NSNumber *) context;
    
    //insert the layer's 'LayerInfo' at the appropriate index of the ToC
    [self insertLayer:layer atIndex:index.intValue];
    
    //remove the observer. 
    [layer removeObserver:self forKeyPath:@"loaded"];
    
    //refresh the flattened tree cache.
    [self.mapViewLevelLayerInfo flattenElementsWithCacheRefresh:YES withLegend:YES];
    
    //reload the table view.
    [self.tableView reloadData];
}


@end

File: TableOfContentsSample/swift/TableOfContents/TOC/Controllers/TOCViewController.swift
//
// Copyright 2015 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

import UIKit
import ArcGIS

protocol TOCViewControllerDelegate:class {
    func dismissTOCViewController(controller:TOCViewController)
}

class TOCViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, LayerInfoCellDelegate {

    @IBOutlet weak var tableView:UITableView!
    
    var itemsArray:[AnyObject]!
    var expandedLayerInfos = [AGSMapContentsLayerInfo]()
    weak var delegate:TOCViewControllerDelegate?
    
    //to hide the status bar for this view controller
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //MARK: - table view data source
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.itemsArray?.count ?? 0
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let LayerInfoCellIdentifier = "LayerInfoCell"
        let LayerLegendCellIdentifier = "LayerLegendCell"
        
        let tempObject: AnyObject = self.itemsArray[indexPath.row]
        
        //if the object is a layer info  then create a LayerInfoCell
        if tempObject is AGSMapContentsLayerInfo {
            let layerInfo = tempObject as! AGSMapContentsLayerInfo
            let layerInfoCell = tableView.dequeueReusableCellWithIdentifier(LayerInfoCellIdentifier) as! LayerInfoCell
            layerInfoCell.level = layerInfo.level
            layerInfoCell.canChangeVisibility = layerInfo.canChangeVisibility
            layerInfoCell.visibility = layerInfo.visible
            layerInfoCell.expanded = layerInfo.expanded
            
            //assign the title.
            layerInfoCell.valueLabel.text = layerInfo.title
            
            //assign the delegate to call the method when the visibility is changed.
            layerInfoCell.layerInfoCellDelegate = self
            
            //no selection style.
            layerInfoCell.selectionStyle = .None
            
            return layerInfoCell
        }
        //else if it is a legend info class, create a legend info cell.
        else {
            let legendElement = tempObject as! AGSMapContentsLegendElement
            let layerLegendCell = tableView.dequeueReusableCellWithIdentifier(LayerLegendCellIdentifier) as! LayerLegendCell
            layerLegendCell.level = legendElement.level
            
            //assign the title.
            layerLegendCell.legendLabel.text = legendElement.title
            
            //assign the image.
            layerLegendCell.legendImage.image = legendElement.swatch
            
            //no selection style.
            layerLegendCell.selectionStyle = .None
            
            return layerLegendCell;
        }
    }
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 44;
    }
    
    //MARK: - table view delegates
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let tempObject: AnyObject = self.itemsArray[indexPath.row]
        
        //if the selected row is of type layer info
        //then expand the layerInfo if not already expanded or vice versa
        if tempObject is AGSMapContentsLayerInfo {
            let layerInfo = tempObject as! AGSMapContentsLayerInfo
            if !layerInfo.expanded {
                //add the children to the items array
                //in order to expand the layer info
                self.addChildrenForSelectedLayerInfo(layerInfo)
                layerInfo.expanded = true
                
//                println("sibling index :: \(self.indexOfNextSibling(layerInfo))")
//                println(self.itemsArray)
            }
            else {
                //else remove the children from the items array
                //iin order to collapse the layer info
                self.removeChildrenForSelectedLayerInfo(layerInfo)
                layerInfo.expanded = false
            }
            //reload the tableview to reflect the changes
            self.tableView.reloadData()
        }
    }
    
    //MARK: - LayerInfoCellDelegate
    
    func layerInfoCell(layerInfoCell: LayerInfoCell, didChangeVisibility visibility: Bool) {
        //retrieve the corresponding cell
        if let cellIndexPath = self.tableView.indexPathForCell(layerInfoCell) {
            //get the layer info represented by the cell
            let layerInfo = self.itemsArray[cellIndexPath.row] as! AGSMapContentsLayerInfo
            
            //set the visibility of the layer info.
            layerInfo.visible = visibility
        }
    }
    
    //MARK: - Helper methods
    
    func addChildrenForSelectedLayerInfo(layerInfo:AGSMapContentsLayerInfo) {
        //add the layer info to be expanded, to the expandedLayerInfos array
        //in order to keep track of all the expanded layer infos
        //we will use this array to reset the state of the layer infos
        //before the view controller is de initialized
        self.expandedLayerInfos.append(layerInfo)
        
        //get the index of the layer info in the items array
        if let index = (self.itemsArray as! [NSObject]).indexOf(layerInfo) {
            //check if the layer info has sublayers
            if layerInfo.subLayers.count > 0 {
                //if true add the sublayers after the parent layer info in the items array
                var i = 1
                for subLayerInfo in layerInfo.subLayers as! [AGSMapContentsLayerInfo] {
                    subLayerInfo.level = layerInfo.level + 1
                    self.itemsArray.insert(subLayerInfo, atIndex: index + i)
                    i++
                }
            }
            //else check if the layer info has legend elements
            else if layerInfo.legendItems.count > 0 {
                //if yes, then add those elements after the parent layer info
                var i = 1
                for legendElement in layerInfo.legendItems as! [AGSMapContentsLegendElement] {
                    legendElement.level = layerInfo.level + 1
                    self.itemsArray.insert(legendElement, atIndex: index + i)
                    i++
                }
            }
        }
    }
    
    func removeChildrenForSelectedLayerInfo(layerInfo:AGSMapContentsLayerInfo) {
        //find the index of the layer info in the expandedLayerInfos array
        //and use the index to remove the layer info from that array,
        //as we will be collapsing this layer info
        if let index = self.expandedLayerInfos.indexOf(layerInfo) {
            self.expandedLayerInfos.removeAtIndex(index)
        }
        
        //find the index of the layer info in the items array
        //and remove the sublayers/legendItems following it
        if let layerInfoIndex = (self.itemsArray as! [NSObject]).indexOf(layerInfo) {
        
            //find the index of the next sibling in the items array
            //remove all the items between the layerInfo index and the sibling layerInfoIndex
            //if sibling not present then simply remove all items after layerInfoIndex
            //also update the expanded state property to false for these items
            var endOfRange:Int = 0
            let siblingLayerInfoIndex = self.indexOfNextSibling(layerInfo)
            if siblingLayerInfoIndex == -2 {
                print("Unexpected error while finding siblings")
                return
            }
            else if siblingLayerInfoIndex == -1 { //no sibling found, simply remove all items after the layerInfo
                endOfRange = self.itemsArray.count-1
            }
            else {
                endOfRange = siblingLayerInfoIndex-1
            }
            
            if layerInfoIndex != endOfRange { //to check for items with no children
                //update the expanded state to false
                for i in layerInfoIndex+1...endOfRange {
                    let tempObject: AnyObject = self.itemsArray[i]
                    if tempObject is AGSMapContentsLayerInfo {
                        (tempObject as! AGSMapContentsLayerInfo).expanded = false
                    }
                }
                //remove the items
                self.itemsArray.removeRange(layerInfoIndex+1...endOfRange)
            }
        }
    }
    
    //find the next sibling in the items array
    //if not present then recursively check for the parent node
    //returns -1 if layerInfo is the last node
    //returns -2 unexpected behavior
    func indexOfNextSibling(layerInfo:AGSMapContentsLayerInfo) -> Int {
        var siblingsArray:[AGSMapContentsLayerInfo]
        //get the siblings array
        if layerInfo.parent != nil {
            siblingsArray = layerInfo.parent!.subLayers as! [AGSMapContentsLayerInfo]
        }
        else {
            //use the root as the parent
            return -1
        }
        
        //find the index of the layerInfo in the siblings array
        if let layerInfoIndex = siblingsArray.indexOf(layerInfo) {
            if layerInfoIndex < siblingsArray.count - 1 {
                //get the sibling layerInfo
                let siblingLayerInfo = siblingsArray[layerInfoIndex+1]
                //find the index of sibling in the itemsArray
                if let siblingLayerInfoIndex = (self.itemsArray as! [NSObject]).indexOf(siblingLayerInfo as NSObject) {
                    return siblingLayerInfoIndex
                }
                else {
                    return -2
                }
            }
            else {
                return self.indexOfNextSibling(layerInfo.parent!)
            }
        }
        return -2
    }
    
    //MARK: - actions
    
    @IBAction func doneAction() {
        self.delegate?.dismissTOCViewController(self)
    }
    
    //MARK: - Deinit
    
    deinit {
        for layerInfo in self.expandedLayerInfos {
            layerInfo.expanded = false
        }
    }
}
